doc/soaptube.1
.\"                                      Hey, EMACS: -*- nroff -*-
.TH SOAPTUBE 1 "April 18, 2005" "Soaptube 0.1.2"


.SH NAME
soaptube \- RTP relay
.\"
.SH SYNOPSIS
.B soaptube
.RI [ option ... ]
.\"
.SH DESCRIPTION
.B soaptube
receives a CD quality WAV RTP stream, and relays it to different media,
icecast2 server or mp3/vorbis files.
It is intended to be used together with
.BR liquidsoap (1),
thus making it independent of output-related failures, like icecast2 crash,
or network lag.
.\"
.SH OPTIONS
.TP
.B \-d, \-\-daemon
Run in daemon mode.
.TP
.BI \-l,\ \-\-load\-conf\-file " file"
Specify the settings file. See
.B HTML
documentation to learn about
.B soaptube
scripts.
.TP
.B \-help, \-\-help
Show summary of options.
.\"
.SH SEE ALSO
.BR liquidsoap (1),
.BR HTML
documentations for settings reference.
.\"
.SH AUTHOR
David Baelde for Savonet <savonet-users@lists.sourceforge.net>


CHANGES
0.3.8 (30-07-2008)
Bugs fixed:
 - Vorbis mono output is now working
 - Fixed parameter parameter description
   in the documentation
 - Propagate new delay in add_timeout
 - Fixed inter-thread mutex lock/unlock in playlist.ml
 - Fixed "next" playlist command
 - Fixed race conditions in request_source.ml feeding task
 - cross/smartcross: raise the default for inhibition as 
   setting it to exactly duration is not enough
 - Don't fail when $HOME is not set
 - Fixed metadata update in input.harbor with icecast2 source protocol
 - Fixed shutdown function. Fixes #153
 - Fixed input.oss. Liquidsoap now works with OSS4 ! Fixes #158
New:
 - Added a hook to execute a function when playlist.once ends
 - Enhanced smart_crossfade
 - Added string.case and string.capitalize
 - New "exec_at" operator, to execute a function depending on a given predicate
 - Added script example in the documentation to listen to radio Nova and get 
   the metadata from a web page.
 - Changed parameters name in fallback.skip to reflect who are the fallback
   and input source.
 - Added a dump file parameter to input.harbor, for debugging purpose.
 - Added an auth function parameter to input.harbor for custom 
   authentifications. Fixes #157
 - Added "primary_queue" and "secondary_queue" commands to request.queue and
   request.equeue sources. Also set the metadata "queue" to either "primary"
   or "secondary" for each request in the queue. Documented it too.
 - Insert latest metadata for a node of a switch source when switching 
   back to it in a middle of a track.
 - Added a 'conservative' parameter to cross, smilar to the one in smartcross.
Internal:
 - Enhanced liqi documentation parser to build the website.

0.3.7 (03-06-2008)
Bugs fixed:
 - Now works on FreeBSD and hopefully other unices that are stricter than
   Linux about Mutex usage.
 - input.http() now has a bind_address parameter.
 - Harbor socket now has a timeout for lost connections.
 - smartcross() is now more compliant with the inter-sources protocol,
   fixes several "get frame didn't change the buffer" bugs.
 - Ogg packeting bugs.
 - Buffering policy in input.http/harbor().
 - No "." in IDs and labels.
 - Resources: FD leaks, useless threads (threads leaks?) in input.http().
 - fade.out() used to run into infinite loops when the delay was 0.
New:
 - New documentation system and website.
 - Self-documenting server with a more helpful "help" command.
 - Moved to duppy: less threads, lighter load, and an configurable scheduler.
 - Moved to Taglib for more reliable access to MP3 metadata.
 - MIME types, notably for playlists and MP3 files.
 - New Jack support. The old one has been renamed to in/output.jack.legacy().
 - Harbor: per-mount passwords and the stop command to kick a source client.
 - Official Last.FM client.
 - Metadata is no more punctual but interval-based, which suppresses some 
   surprising behaviours.
 - Perfected daemon behaviour.
 - All output.file.*() now have the features that used to be only in 
   output.file.vorbis(), notable re-opening. Added %w to the strftime-like 
   format codes allowed in their filename parameter.
 - Add clear_metadata() and map_metadata(). Now, rewrite_metadata() is a simple
   specialization of map_metadata(), written in utils.liq.
 - Dynamic amplification factor in amplify(), e.g. useful for replay gain.
 - Lots of new functions in the scripting API: for lists, requests, system 
   interaction, shutdown, command-line parsing, scripted server commands, etc.
As always:
 - code cleanup, style, etc.

0.3.6 (17-12-2007)
Bugfix release:
 - Close Http socket
 - Add http socket timeout
 - Close playlist file after reading its content
 - Fix http redirect support for lastfm files
 - Fix file open leak in camomile support
 - Fix playlist uri ending with "/"

0.3.5 (08-11-2007)
Bugfix release:
  - Fixed #57: scpls and mpegurl playlist parsing
  - Fixed #46: Late cross-scripts bindings

0.3.4 (25-09-2007)
Notation: "-" stands for a change, "+" for an addition.
* Language
  - Support for polymorphism, subtyping and basic ad-hoc polymorphism,
    which allows a much simpler API, notably for maths and serialization.
  + Added interactive_*() for mutable values.
  - The right syntax for settings is now set("var", value) and can be used
    anywhere in the scripts.
  - The volume parameters of most operators are now in dB.
  - Many builtin functions added.
  - Nicer type error messages.
* Sources
  + Added input.lastfm() to relay last.fm streams.
  + Added input.harbor() to received Icecast2 source streams.
  + Added noise() to generate white noise
  - Reimplemented playlist support, added various xml and text formats.
  - Added mpd protocol to find files using mpd.
* Operators
  + New effects: compress(), flanger(), pan().
  + New filters: filter.fir.*(), filter.iir.*(), filter.biquad.*(), comb().
  + Added support for LADSPA effects.
  + Added eat_blank() to remove blanks.
* Outputs
  - Added non-default restart option for output.icecast.*().
  - Added the possibility to tweak some settings at runtime.
  - Split output.icecast.vorbis() into output.icecast.vorbis.*() to distinguish
    between encoding modes -- and similarly for output.file.vorbis and mp3.
  - Better handling of Icecast disconnections.
* IO
  + Added portaudio support.
  - Jack support is now somewhat working.
* As usual, lots of bug fixes, careful polishing & much more...

0.3.3 (06-06-2007)
* Major cleanup of the core stream representation; moved to float arrays,
  removing several back-and-forth conversions and enhancing the perfs a lot;
  reviewed all sources and operators, made many minor enhancements btw.
* Lots of sound processing operators: compand, compress, normalize,
  pitch, bpm, soundtouch, saw, square, etc. Add more shapes to fade.*().
* New track processing operators: insert_metadata, on_track.
* Smart cross: allows to select a transition based on the volumes around the 
  end-of-track.
* Support for AAC encoding/decoding.

src/outputs/soaptube.ml

open Dtools

(*<section soaptube>*)

let log = Log.log ~label:"soaptube"

let usage =
  "\n" ^
  "Copyright (c) 2003 Savonet Team.\n" ^
  "This software is under GNU GPL v2.\n" ^
  "Read COPYING for more information.\n" ^
  "\n" ^
  "Liquidrelay relays liquidsoap output.\n" ^
  "Liquidrelay is part of the Savonet project : <http://savonet.sf.net>.\n\n" ^
  "Usage : liquidsoap [options]\n" ^
  "        where options are ...\n"

let input = ref ""

let options =
  List.fold_left
    ( fun l (la,b,c) ->
        let ta = List.hd (List.rev la) in
        let expand = List.map
                       (fun a -> (a,b,(if a = ta then ("\n\t  "^c) else "")))
                       la in
          l@expand ) []
    [

      ["-d";"--daemon"],
      Arg.Unit (fun f -> Conf.set_bool "daemon" true),
      "Run in daemon mode" ;

      ["-l";"--load-conf-file"],
      Arg.String (fun f -> input := f),
      "Configuration file" ;

    ]

let anon_fun = fun s -> Printf.fprintf stderr "Unused argument '%s'\n" s

let ignore_sigpipe =
  Sys.set_signal Sys.sigpipe Sys.Signal_ignore

let read_conf f =
  let basename = Filename.basename f in
  let basename =
    try Filename.chop_extension basename with _ -> basename
  in
  let default_log = basename ^ ".log" in
  let default_pid = basename ^ ".pid" in

    Dtools.Conf.set_string "log.file"
      (*<ignore>*)
      ( let i = Dtools.Conf.get_string "log.file" in
          (* log.file is a builtin key, default is "" *)
          if i = "" then default_log else i ) ;

    Dtools.Conf.set_string "daemon.pidfile"
      (*<ignore>*)
      ( let i = Dtools.Conf.get_string "daemon.pidfile" in
          (* It is also a builtin key, default is "" *)
          if i = "" then default_pid else i ) ;

    Dtools.Conf.set_string "log.dir"
      (* The log file will be <log.dir>/<log.file>. *) 
      (Dtools.Conf.get_string ~default:Configure.logdir "log.dir") ;
    Dtools.Conf.set_string "log.file"
      ( (*<ignore>*)
	(Dtools.Conf.get_string "log.dir")^"/"^
	(* Default is set to <name>.log if radio program is <name>.liq *)
	(Dtools.Conf.get_string "log.file")) ;

    Dtools.Conf.set_string "daemon.piddir"
      (* The pid file will be <daemon.piddir>/<daemon.pidfile> *) 
      (Dtools.Conf.get_string ~default:Configure.piddir "daemon.piddir"); 
    Dtools.Conf.set_string "daemon.pidfile"
      ( (*<ignore>*)
	(Dtools.Conf.get_string "daemon.piddir")^"/"^
	(* Default is set to <name>.pid if radio program is <name>.liq *)
	(Dtools.Conf.get_string "daemon.pidfile")) ;

    try Dtools.Conf.read_file f with _ ->
      failwith (Printf.sprintf "Cannot read conf file %S !" f)

let log = Log.log ~label:"output"

(* Where to get the RTP stream *)
let ip = Dtools.Conf.get_string ~default:"224.0.1.20" "rtp.address"
let port = Dtools.Conf.get_int ~default:8888 "rtp.port"
let new_session () = Rtp.new_session Rtp.Recv ip port
let session = ref (new_session ())
let wav = Mixer.Buffer.create ()

let loop f =
  let cr = Mutex.create () in
  let cw = Mutex.create () in

  let encoder () =
    let cp =
      let b = Mixer.Buffer.create () in
        Mixer.Buffer.set_already b Mixer.Buffer.size ;
        b
    in
      while true do
        Mutex.lock cr ;

        String.blit
          (Mixer.Buffer.to_string wav) 0 (Mixer.Buffer.to_string cp) 0
          Mixer.Buffer.size ;
        Mixer.Buffer.free_metadatas cp ;
        begin match Mixer.Buffer.get_metadata wav with
          | None -> ()
          | Some m -> Mixer.Buffer.push_metadata cp m
        end ;

        Mutex.unlock cw ;
        f cp
      done
  in
    Mutex.lock cr ;
    ignore (Thread.create encoder ()) ;
    while true do
      Mutex.lock cw ;

      if Rtp.recv !session wav then
        ( log 2 "Garbage on the session: rebooting it!" ;
          session := new_session () ) ;

      Mutex.unlock cr
    done

let _ =
  Arg.parse options anon_fun usage ;
  read_conf !input ;
  let output = Output.from_conf () in
    Init.init
      (fun () ->
         while (Mixer.Buffer.to_string wav).[0] = '\000' do
           if Rtp.recv !session wav then
             (Mixer.Buffer.to_string wav).[0] <- '\000'
         done ;
         log 2 "Stream started!" ;
         output loop )
